pkg/telnetlib/connection.go

```diff
@@ -0,0 +1,297 @@
+// databuffer is written to by the FSM and it is read from by the dataHandler
+type dataReadWriter struct {
+	dataBuffer bytes.Buffer
+	dataMux    *sync.Mutex
+}
+
+func (drw *dataReadWriter) Read(p []byte) (int, error) {
+	drw.dataMux.Lock()
+	defer drw.dataMux.Unlock()
+	return drw.dataBuffer.Read(p)
+}
+
+func (drw *dataReadWriter) Write(p []byte) (int, error) {
+	drw.dataMux.Lock()
+	defer drw.dataMux.Unlock()
+	return drw.dataBuffer.Write(p)
+}
+
+type connectionWriter struct {
+	ch chan []byte
+}
+
+func (cw *connectionWriter) Write(b []byte) (int, error) {
+	if cw.ch != nil {
+		cw.ch <- b
+	}
+	return len(b), nil
+}
+
+func (cw *connectionWriter) Close() error {
+	close(cw.ch)
+	cw.ch = nil
+	return nil
+}
+
+func newTelnetConn(opts connOpts) *TelnetConn {
+	tc := &TelnetConn{
+		conn:               opts.conn,
+		writeCh:            make(chan []byte),
+		unackedServerOpts:  make(map[byte]bool),
+		unackedClientOpts:  make(map[byte]bool),
+		cmdHandler:         opts.cmdHandler,
+		dataHandler:        opts.dataHandler,
+		dataHandlerCloseCh: make(chan chan struct{}),
+		dataWrittenCh:      make(chan bool),
+		serverOpts:         opts.serverOpts,
+		clientOpts:         opts.clientOpts,
+		optionCallback:     opts.optCallback,
+		connWriteDoneCh:    make(chan chan struct{}),
+		negotiationDone:    make(chan struct{}),
+		closed:             false,
+	}
+	if tc.optionCallback == nil {
+		tc.optionCallback = tc.handleOptionCommand
+	}
+	tc.handlerWriter = &connectionWriter{
+		ch: tc.writeCh,
+	}
+	tc.dataRW = &dataReadWriter{
+		dataMux: new(sync.Mutex),
+	}
+	fsm := opts.fsm
+	fsm.tc = tc
+	tc.fsm = fsm
+	for k := range tc.serverOpts {
+		tc.unackedServerOpts[k] = true
+	}
+	for k := range tc.clientOpts {
+		tc.unackedClientOpts[k] = true
+	}
+	return tc
+}
+
+func (c *TelnetConn) writeLoop() {
+	log.Debugf("entered write loop")
+	for {
+		select {
+		case writeBytes := <-c.writeCh:
+			c.conn.Write(writeBytes)
+		case ch := <-c.connWriteDoneCh:
+			ch <- struct{}{}
+			return
+		}
+	}
+}
+
+func (c *TelnetConn) startNegotiation() {
+	c.srvrOptsLock.Lock()
+	for k := range c.serverOpts {
+		log.Infof("sending WILL %d", k)
+		c.unackedServerOpts[k] = true
+		c.sendCmd(Will, k)
+	}
+	c.srvrOptsLock.Unlock()
+	c.clientOptsLock.Lock()
+	for k := range c.clientOpts {
+		log.Infof("sending DO %d", k)
+		c.unackedClientOpts[k] = true
+		c.sendCmd(Do, k)
+	}
+	c.clientOptsLock.Unlock()
+	select {
+	case <-c.negotiationDone:
+		log.Infof("Negotiation finished")
+		return
+	case <-time.After(10 * time.Second):
+		log.Infof("Negotiation failed. Exiting")
+		c.Close()
+		c.closed = true
+		return
+	}
+}
+
+// Close closes the telnet connection
+func (c *TelnetConn) Close() {
+	log.Infof("Closing the connection")
+	c.conn.Close()
+	c.closeConnLoopWrite()
+	c.closeDatahandler()
+	c.handlerWriter.Close()
+	log.Infof("telnet connection closed")
+	c.closedMutex.Lock()
+	defer c.closedMutex.Unlock()
+	c.closed = true
```

Comment:
just curious, why is it is OK to call c.conn.Close/closeConnLoopWrite/closeConnLoopWrite and closeConnLoopWrite without taking the lock but setting c.closed is not?